clang-tools-extra/clang-tidy/bugprone/NullCheckAfterDereferenceCheck.cpp

```diff
@@ -0,0 +1,125 @@
+//===--- NullCheckAfterDereferenceCheck.cpp - clang-tidy
+//--------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "NullCheckAfterDereferenceCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/DeclCXX.h"
+#include "clang/AST/DeclTemplate.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+#include "clang/Analysis/CFG.h"
+#include "clang/Analysis/FlowSensitive/ControlFlowContext.h"
+#include "clang/Analysis/FlowSensitive/DataflowAnalysisContext.h"
+#include "clang/Analysis/FlowSensitive/DataflowEnvironment.h"
+#include "clang/Analysis/FlowSensitive/DataflowLattice.h"
+#include "clang/Analysis/FlowSensitive/Models/NullCheckAfterDereferenceModel.h"
+#include "clang/Analysis/FlowSensitive/WatchedLiteralsSolver.h"
+#include "clang/Basic/SourceLocation.h"
+#include "llvm/ADT/Any.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/Error.h"
+#include <memory>
+#include <vector>
+
+namespace clang::tidy::bugprone {
+
+using ast_matchers::MatchFinder;
+using dataflow::NullCheckAfterDereferenceModel;
+using dataflow::ReverseNullDiagnoser;
+
+static constexpr llvm::StringLiteral FuncID("fun");
+
+static std::optional<ReverseNullDiagnoser::ResultType>
+analyzeFunction(const FunctionDecl &FuncDecl, ASTContext &ASTCtx) {
```

Comment:
✏️ A small code readability optimisation opportunity here: if you have a valid `Decl` in your hand, AFAIK, you can always get to its valid `ASTContext`, so passing both parameters this explicitly is not needed.

```suggestion
analyzeFunction(const FunctionDecl &FD) {
  const ASTContext& Ctx = FD.getContext();
```

Reply:
Indeed, changed to `getASTContext()`.